### ISSUE TITLE
Strategy for managing computational complexity

### ISSUE THREAD

**Issue by HeinrichApfelmus** [DeltaQ-SD/deltaq#72]
## Problem

The section [Performance measurements](https://github.com/DeltaQ-SD/dq-revamp/blob/41e75562843dcad0eae2d10fc678f7efb7c11a17/doc/reports/artefact-A2.md#performance-measurements) of the [Report on test results, coverage analysis and performance](https://github.com/DeltaQ-SD/dq-revamp/blob/41e75562843dcad0eae2d10fc678f7efb7c11a17/doc/reports/artefact-A2.md) highlights that repeated application of sequential composition, `.>>.`, creates probability distributions with large internal representation size. (The internal representation size is measured by the function `complexity`.)

Specifically, sequentially composition of $N$ outcomes, `o1 .>>. o2 .>>. … .>>. on`, will give a probability distribution whose internal representation size is **exponential** in $N$. This means that computing the result will become exponentially slow as the number of outcomes increases. As such compositions represent typical uses cases, this slowdown is a **problem for usability**.

Unfortunately, the problem is inherent in the **semantics** of **sequential composition**. For instance, consider statistical mixtures of discrete Dirac measures at the powers of two,

```hs
import Numeric.Measure.Discrete
import qualified Data.Map as Map

mix :: Int → Discrete Rational
mix m = scale 0.5 $ add (dirac 0) (dirac (2^(m :: Int)))
```

Then, the `n`-fold convolution

```hs
convolved :: Int → Discrete Rational
convolved n = foldr1 convolve $ map mix [1..n]
```

will have internal representation size `2^n`. For example

```
>>> Map.size $ toMap $ convolved 10
1024
```

## Solution Proposals 

The problem statement indicates that an exact representation of probability distributions will invariably lead to exponentially growing internal representation size. In order to curb this growth, any solution is **forced** to **approximate**.

### Common requirements

In order to allow the user to reason about approximate results, it's sensible to keep the exact representation, but give them choice by providing a function

```hs
approximate :: DeltaQ o => Rational -> o -> o
```

such that `approximate eps x` returns a result with reduced internal size that is within a precision `eps` in some appropriate metric.

Furthermore, we expect that `approximate` is conservative, that is it only adds delay.

```hs
∀ t eps x. successWithin t (approximate eps x) <= successWithin t x
```

For convenience, a function `approx = approximate eps'` with sensible default precision `eps'` should also be provided.

### Proposal 1, "Chebychev"

We could switch the representation of polynomials to a different basis. The basis of Chebychev polynomials offers favorable conditions for approximation. However, as the problem statement discusses, the issue of exponential growth of the representation size already applies for discrete probability distributions, so using a different presentation for polynomials will have little effect.

### Proposal 2, "Discretize probability"

Approximate in probability, i.e. consider the cumulative probabilities of `44.999 %` and `50.000 %` approximately equal. In other words, the idea is to approximate `\p -> quantile p x` by a step function with few steps.

This requires more thought:

- [ ] Which metric does this approximate?
- [ ] Error estimates for the operations `.>>.`, `./\.`, `.\/.`, `choice` under this metric?
- [ ] Log-scale for probability, in order to better discern orders of magnitude of failure?
- [ ] Sensible default precision and scale. Do we need failure probabilities beyond $10^{-7}$?

**Comment by peterwthompson** [comment]
This is resolved by PR #92.

**Comment by peterwthompson** [comment]
This is resolved by PR https://github.com/DeltaQ-SD/deltaq/pull/92.
